### PR TITLE
Test/diagonal win

### PR DESCRIPTION
**Pull request by bkeener7**
This PR will add all completed win conditions with their corresponding tests to the main branch. By using the `connect_four` method, all four win conditions will be checked with one command from the turn.rb file.

### CODE REVIEW

File: lib/turn.rb [hunk 1]
```diff
@@ -54,34 +54,59 @@ def row_win
...
-    arr1 = [board.columns[0].rows[3], board.columns[1].rows[2], board.columns[2].rows[1], board.columns[3].rows[0]]
-    arr2 = [board.columns[0].rows[4], board.columns[1].rows[3], board.columns[2].rows[2], board.columns[3].rows[1], board.columns[4].rows[0]]
-    arr3 = [board.columns[0].rows[5], board.columns[1].rows[4], board.columns[2].rows[3], board.columns[3].rows[2], board.columns[4].rows[1], board.columns[5].rows[0]]
-    arr4 = [board.columns[1].rows[5], board.columns[2].rows[4], board.columns[3].rows[3], board.columns[4].rows[2], board.columns[5].rows[1], board.columns[6].rows[0]]
-    arr5 = [board.columns[2].rows[5], board.columns[3].rows[4], board.columns[4].rows[3], board.columns[5].rows[2], board.columns[6].rows[1]]
-    arr6 = [board.columns[3].rows[5], board.columns[4].rows[4], board.columns[5].rows[3], board.columns[6].rows[2]]
-
-    right_up = [arr1, arr2, arr3, arr4, arr5, arr6]
-
+    arr0 = [board.columns[0].rows[3].player, board.columns[1].rows[2].player, board.columns[2].rows[1].player, board.columns[3].rows[0].player]
```
Comment (msakr21) [inline lib/turn.rb:57]: Great change to easily recognize which array element these are! I love how you implemented what we discussed on our visual diagram. I'll attach it here for our assessor's reference.

The different arrays follow the different red, orange and yellow diagonal set ups. The black dots were what we had determined are essential plays for the different possible diagonal wins.  We focused on the diagonal in one direction from bottom left to top right, initially.

8860eab3-3682-404c-a830-e5a20b0cdc7a

File: lib/turn.rb [hunk 2]
```diff
@@ -54,34 +54,59 @@ def row_win
...
+    arr1 = [board.columns[0].rows[4].player, board.columns[1].rows[3].player, board.columns[2].rows[2].player, board.columns[3].rows[1].player, board.columns[4].rows[0].player]
+    arr2 = [board.columns[0].rows[5].player, board.columns[1].rows[4].player, board.columns[2].rows[3].player, board.columns[3].rows[2].player, board.columns[4].rows[1].player, board.columns[5].rows[0].player]
+    arr3 = [board.columns[1].rows[5].player, board.columns[2].rows[4].player, board.columns[3].rows[3].player, board.columns[4].rows[2].player, board.columns[5].rows[1].player, board.columns[6].rows[0].player]
+    arr4 = [board.columns[2].rows[5].player, board.columns[3].rows[4].player, board.columns[4].rows[3].player, board.columns[5].rows[2].player, board.columns[6].rows[1].player]
+    arr5 = [board.columns[3].rows[5].player, board.columns[4].rows[4].player, board.columns[5].rows[3].player, board.columns[6].rows[2].player]
+    right_up = [arr0, arr1, arr2, arr3, arr4, arr5]
```
Comment (msakr21) [inline lib/turn.rb:63]: This is much clearer than what I'd seen earlier in the day!

File: lib/turn.rb [hunk 3]
```diff
@@ -54,34 +54,59 @@ def row_win
...
+    
+    #key checks are in arr0[0], arr1[1], arr2[2], arr3[2], arr4[1], arr5[0]
```
Comment (msakr21) [inline lib/turn.rb:65]: I like how you put a comment on the key checks I mentioned in my earlier comment here! Helps with following the code!

File: lib/turn.rb [hunk 4]
```diff
@@ -54,34 +54,59 @@ def row_win
...
+    check_right = [right_up[0][0], right_up[1][1], right_up[2][2], right_up[3][2], right_up[4][1], right_up[5][0]]
+    
+  
+   if check_right.any? { |occupied| occupied.size > 0 }
+      right_up.each do |arr|
+        if arr.chunk_while{ |a, b| a == b && a != "" }.any?{ |player_piece| player_piece.size == 4}       
+          return arr.max_by{ |player| arr.count(player)}
+        end             
+      end    
+    end
```
Comment (msakr21) [inline lib/turn.rb:75]: This seems like it needs a space removed.

File: lib/turn.rb [hunk 5]
```diff
@@ -54,34 +54,59 @@ def row_win
...
-
-    left_up = [arr1, arr2, arr3, arr4, arr5, arr6]  
-
+    arr0 = [board.columns[3].rows[5].player, board.columns[2].rows[4].player, board.columns[1].rows[3].player, board.columns[0].rows[2].player]
+    arr1 = [board.columns[4].rows[5].player, board.columns[3].rows[4].player, board.columns[2].rows[3].player, board.columns[1].rows[2].player, board.columns[0].rows[1].player]
+    arr2 = [board.columns[5].rows[5].player, board.columns[4].rows[4].player, board.columns[3].rows[3].player, board.columns[2].rows[2].player, board.columns[1].rows[1].player, board.columns[0].rows[0].player]
+    arr3 = [board.columns[6].rows[5].player, board.columns[5].rows[4].player, board.columns[4].rows[3].player, board.columns[3].rows[2].player, board.columns[2].rows[1].player, board.columns[1].rows[0].player]
+    arr4 = [board.columns[6].rows[4].player, board.columns[5].rows[3].player, board.columns[4].rows[2].player, board.columns[3].rows[1].player, board.columns[2].rows[0].player]
+    arr5 = [board.columns[6].rows[3].player, board.columns[5].rows[2].player, board.columns[4].rows[1].player, board.columns[3].rows[0].player]
+    left_up = [arr0, arr1, arr2, arr3, arr4, arr5]
```
Comment (msakr21) [inline lib/turn.rb:87]: For assessor or viewer: this is where we check the diagonal going top left down to bottom right of the board.

**Review by msakr21 — COMMENTED** [review]
This looks very cleanly executed! Excellent job!

**Review by msakr21 — COMMENTED** [review]
Is there a reason for the removal of the @ from the player and board attributes on lines 12, 22 and 42?

**Review by msakr21 — COMMENTED** [review]
These look good at a cursory glance. I'll go ahead and merge to main and if we end up needing to change anything later on, we'll go over the changes together!

**Comment by msakr21** [comment]
I'll go ahead and merge these changes into main. Please check my comment on lines 12, 22 and 42 just so I'd learn/understand the reasoning behind that change.

Really happy with the addition to our code! Great job, Bryan!

File: lib/turn.rb (hunk 1 above)
Comment (bkeener7) [reply lib/turn.rb:57]: Very helpful visual guide indeed, I'm glad you sent this as it's what ultimately made everything *click* for the diagonal win methods.

File: lib/turn.rb (hunk 3 above)
Comment (bkeener7) [reply lib/turn.rb:65]: I'm working on documenting more frequently!

File: lib/turn.rb (hunk 4 above)
Comment (bkeener7) [reply lib/turn.rb:75]: I've been trying to squash little inconsistencies like this whenever I see them. Once we have the final code, I will do one last sweep-through and get everything to be visually identical.

**Comment by bkeener7** [comment]
> Is there a reason for the removal of the @ from the player and board attributes on lines 12, 22 and 42?

Mostly consistency, especially for board. There are around almost 70 `board` calls in the turn.rb file and having the @ made the `diagonal_rightup` and `diagonal_leftup` methods almost unreadable.

**Pull request merged**